storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java

```diff
@@ -1297,6 +1390,59 @@ private void removeNonCanonicalBlobSidecars(
     }
   }
 
+  private void removeNonCanonicalSidecars(
+      final Map<Bytes32, UInt64> deletedHotBlocks,
+      final Map<Bytes32, Bytes32> finalizedChildToParentMap) {
+
+    final Set<SlotAndBlockRoot> nonCanonicalBlocks =
+        deletedHotBlocks.entrySet().stream()
+            .filter(entry -> !finalizedChildToParentMap.containsKey(entry.getKey()))
+            .map(entry -> new SlotAndBlockRoot(entry.getValue(), entry.getKey()))
+            .collect(Collectors.toSet());
+
+    if (storeNonCanonicalBlocks) {
+      final Iterator<SlotAndBlockRoot> nonCanonicalBlocksIterator = nonCanonicalBlocks.iterator();
+      int index = 0;
+      while (nonCanonicalBlocksIterator.hasNext()) {
+        final int start = index;
+        try (final FinalizedUpdater updater = finalizedUpdater()) {
+          while (nonCanonicalBlocksIterator.hasNext() && (index - start) < BLOBS_TX_BATCH_SIZE) {
+            dao.getDataColumnIdentifiers(nonCanonicalBlocksIterator.next())
+                .forEach(
+                    key -> {
+                      dao.getSidecar(key)
+                          .ifPresent(
+                              sidecarBytes -> {
+                                DataColumnSidecar sideCar =
+                                    spec.deserializeSidecar(sidecarBytes, key.slot());
+                                updater.addNonCanonicalSidecar(sideCar);
+                                LOG.trace(
+                                    "Moving non-canonical sidecar with identifier {} to non-canonical sidecars table",
+                                    key);
+                                updater.removeSidecar(key);
```

Comment:
my assumption is this isnt tested much, given its complexity. we should raise a testing and/or refactor issue for this

Reply:
We'll have a teku-internal ticket all about testing 👍

Reply:
Created https://github.com/Consensys/teku/issues/9437 to track this.